src/coprocessor/endpoint.rs

```diff
@@ -185,29 +186,31 @@ impl Host {
                 pool.spawn(move |_| on_resp.respond_stream(s)).forget();
             }
             CopRequest::Analyze(analyze) => {
-                let ctx = AnalyzeContext::new(analyze, ranges, snap, &req_ctx);
-                let mut exec_metrics = ExecutorMetrics::default();
+                let mut ctx = AnalyzeContext::new(analyze, ranges, snap, &req_ctx).unwrap();
                 let do_request = move |_| {
                     tracker.record_wait();
-                    let mut resp = ctx.handle_request(&mut exec_metrics).unwrap_or_else(|e| {
+                    let mut resp = ctx.handle_request().unwrap_or_else(|e| {
                         let mut metrics = tracker.get_basic_metrics();
                         err_resp(e, &mut metrics)
                     });
+                    let mut exec_metrics = ExecutorMetrics::default();
+                    ctx.collect_metrics_into(&mut exec_metrics);
                     tracker.record_handle(Some(&mut resp), exec_metrics);
                     on_resp.respond(resp);
                     future::ok::<_, ()>(())
                 };
                 pool.spawn(do_request).forget();
             }
             CopRequest::Checksum(checksum) => {
-                let ctx = ChecksumContext::new(checksum, ranges, snap, &req_ctx);
-                let mut exec_metrics = ExecutorMetrics::default();
+                let mut ctx = ChecksumContext::new(checksum, ranges, snap, &req_ctx).unwrap();
                 let do_request = move |_| {
                     tracker.record_wait();
-                    let mut resp = ctx.handle_request(&mut exec_metrics).unwrap_or_else(|e| {
+                    let mut resp = ctx.handle_request().unwrap_or_else(|e| {
                         let mut metrics = tracker.get_basic_metrics();
                         err_resp(e, &mut metrics)
                     });
+                    let mut exec_metrics = ExecutorMetrics::default();
+                    ctx.collect_metrics_into(&mut exec_metrics);
```

Comment:
Seems L206~218 repeated in all Unary request, could we merge it now?

Reply:
that's exactly what we will do in #2896